src/swarm/neo/util/Batch.d

```diff
@@ -148,9 +148,78 @@ struct BatchWriter
     }
 }
 
-struct BatchReader
+scope class BatchReader ( T ... )
 {
+    private Const!(void)[] batch;
+    private Const!(void)[] remaining;
+
+    public this ( in void[] batch )
+    {
+        this.batch = batch;
+        this.remaining = this.batch;
+    }
+
+    // compress_buf must remain accessible for the life of this instance
+    public this ( Lzo lzo, in void[] batch, ref void[] compress_buf )
+    {
+        // Read uncompressed length from first size_t.sizeof bytes.
+        auto uncompressed_len = *(cast(size_t*)(batch.ptr));
+        compress_buf.length = uncompressed_len;
+
+        // Decompress into this.batch.
+        auto src = batch[size_t.sizeof .. $];
+        lzo.uncompress(src, compress_buf);
 
+        // Construct as normal, extracting from the decompressed data.
+        this(compress_buf);
+    }
+
+    public int opApply ( int delegate ( ref T ) dg )
+    {
+        T items;
+        while ( this.remaining.length )
+        {
+            foreach ( ref item; items )
+                this.extractItem(item);
+
+            if ( auto r = dg(items) )
+                return r;
+        }
+
+        return 0;
+    }
+
+    private void extractItem ( T ) ( ref T item )
```

Comment:
Hmm, I would argue this should be `public` and probably `opApply` removed. The reason is that you can pack key and value of the distinct types, but you can't extract them with the `opApply`.

Reply:
Yes you can. See the unittest.

Reply:
If you're referring to the last unittest, that's just packing a struct of the `struct S { key; value; }` I was thinking about this:

Node:
```
time_t key; void[] value;
batch.add(key, value);
```

How you extract this? Because, you can pack it.

Reply:
It looks a bit deceptive. Look at the lines where it [adds to the batch](https://github.com/sociomantic-tsunami/swarm/pull/188/files#diff-f9e677247e1ecb227ec79f3aebc4a829R379) and [extracts from the batch](https://github.com/sociomantic-tsunami/swarm/pull/188/files#diff-f9e677247e1ecb227ec79f3aebc4a829R384).

Reply:
(The struct is only used as a convenient way of making a list of "records" to be put in the batch.)

Reply:
Hmm, all right. Though I think it looks weird, but I'll not complain. Would you be against adding `public bool extractItems(T...)(ref T items)`?

Reply:
>Would you be against adding public bool extractItems(T...)(ref T items)?

To extract just a single tuple / record? I wondered if there's a use case for that.

>Though I think it looks weird, but I'll not complain.

If you think the API looks weird, please complain. If you think the unittest looks weird, I'm not so bothered.

Reply:
I think I'm missing something here. Can you use `opApply` to extract the tuple or arbitrary degree?

> To extract just a single tuple / record?

To extract `(a1, a2, a3, ..., an)` (same as you can add them using `Batch.add` method)

Reply:
Yes, I guess you're missing something ;) This is what the usage looks like:

Node-side:
```D
hash_t key;
mstring value;
batch.add(key, value);
```
Client-side:
```D
foreach ( key, value; new BatchReader!(hash_t, cstring)(batch) )
```

Reply:
For the record, I was missing the template `BatchReader` argument.

Reply:
Suggest changing the name of the method template parameter to `Item` or something to avoid confusion with the class template parameter.